### PR TITLE
gh-111789: Use PyDict_GetItemRef() in Objects/

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* Issue: gh-111789

### CODE REVIEW

**Comment by vstinner** [comment]
Ping me when this PR is ready for a review. Now it's marked as a draft.

**Comment by serhiy-storchaka** [comment]
It is ready for review. I marked is as draft because I do not insist that all these changes should be applied.

File: Objects/object.c [hunk 1]
```diff
@@ -1488,19 +1488,14 @@ _PyObject_GetMethod(PyObject *obj, PyObject *name, PyObject **method)
     }
     if (dict != NULL) {
         Py_INCREF(dict);
-        PyObject *attr = PyDict_GetItemWithError(dict, name);
-        if (attr != NULL) {
-            *method = Py_NewRef(attr);
+        if (PyDict_GetItemRef(dict, name, method) != 0) {
+            // error or found
```
Comment (serhiy-storchaka) [inline Objects/object.c]:
```suggestion
            // found or error
```

File: Objects/typeobject.c [hunk 2]
```diff
@@ -10263,23 +10240,18 @@ _super_lookup_descr(PyTypeObject *su_type, PyTypeObject *su_obj_type, PyObject *
...
     do {
         PyObject *obj = PyTuple_GET_ITEM(mro, i);
         PyObject *dict = lookup_tp_dict(_PyType_CAST(obj));
         assert(dict != NULL && PyDict_Check(dict));
 
-        res = PyDict_GetItemWithError(dict, name);
-        if (res != NULL) {
-            Py_INCREF(res);
+        if (PyDict_GetItemRef(dict, name, &res) != 0) {
+            // error or found
```
Comment (serhiy-storchaka) [inline Objects/typeobject.c]:
```suggestion
            // found or error
```

File: Objects/object.c [hunk 3]
```diff
@@ -1608,20 +1603,17 @@ _PyObject_GenericGetAttrWithDict(PyObject *obj, PyObject *name,
     if (dict != NULL) {
         Py_INCREF(dict);
         res = PyDict_GetItemWithError(dict, name);
```
Comment (vstinner) [inline Objects/object.c]: Did you forget to remove this call?

File: Objects/typeobject.c [hunk 4]
```diff
@@ -1435,18 +1429,12 @@ type_get_doc(PyTypeObject *type, void *context)
...
-                                               (PyObject *)type);
+    if (PyDict_GetItemRef(dict, &_Py_ID(__doc__), &result) == 0) {
+        result = Py_NewRef(Py_None);
     }
-    else {
-        Py_INCREF(result);
+    else if (result && Py_TYPE(result)->tp_descr_get) {
+        Py_SETREF(result, Py_TYPE(result)->tp_descr_get(result, NULL,
+                                               (PyObject *)type));
     }
```
Comment (vstinner) [inline Objects/typeobject.c]:
```suggestion
    else if (result) {
        descrgetfunc descr_get = Py_TYPE(result)->tp_descr_get;
        if (descr_get) {
            Py_SETREF(result, descr_get(result, NULL, (PyObject *)type));
        }
    }
```

File: Objects/typeobject.c [hunk 5]
```diff
@@ -1477,16 +1465,16 @@ type_get_annotations(PyTypeObject *type, void *context)
...
+    }
     if (annotations) {
         if (Py_TYPE(annotations)->tp_descr_get) {
-            annotations = Py_TYPE(annotations)->tp_descr_get(
-                    annotations, NULL, (PyObject *)type);
-        } else {
-            Py_INCREF(annotations);
+            Py_SETREF(annotations, Py_TYPE(annotations)->tp_descr_get(
+                    annotations, NULL, (PyObject *)type));
         }
```
Comment (vstinner) [inline Objects/typeobject.c]: Here I suggest:

```c
        descrgetfunc get = Py_TYPE(annotations)->tp_descr_get;
        if (get) {
            Py_SETREF(annotations, get(annotations, NULL, (PyObject*)type));
        }
```

File: Objects/typeobject.c [hunk 6]
```diff
@@ -1533,16 +1521,11 @@ type_set_annotations(PyTypeObject *type, PyObject *value, void *context)
...
-    PyObject *params = PyDict_GetItemWithError(lookup_tp_dict(type), &_Py_ID(__type_params__));
-
-    if (params) {
-        return Py_NewRef(params);
-    }
-    if (PyErr_Occurred()) {
-        return NULL;
+    PyObject *params;
+    if (PyDict_GetItemRef(lookup_tp_dict(type), &_Py_ID(__type_params__), &params) == 0) {
+        params =  PyTuple_New(0);
```
Comment (vstinner) [inline Objects/typeobject.c]:
```suggestion
        return  PyTuple_New(0);
```

**Review by vstinner — DISMISSED** [review]
LGTM with some suggestions

**Comment by vstinner** [comment]
> LGTM with some suggestions

Sorry, _PyObject_GenericGetAttrWithDict() should be fixed before I can approve the PR.

File: Objects/object.c (hunk 3 above)
Comment (serhiy-storchaka) [inline Objects/object.c]:
```suggestion
```

Comment (serhiy-storchaka) [reply Objects/object.c]: Good catch!

File: Objects/typeobject.c [hunk 7]
```diff
@@ -1533,16 +1523,11 @@ type_set_annotations(PyTypeObject *type, PyObject *value, void *context)
...
-    PyObject *params = PyDict_GetItemWithError(lookup_tp_dict(type), &_Py_ID(__type_params__));
-
-    if (params) {
-        return Py_NewRef(params);
-    }
-    if (PyErr_Occurred()) {
-        return NULL;
+    PyObject *params;
+    if (PyDict_GetItemRef(lookup_tp_dict(type), &_Py_ID(__type_params__), &params) == 0) {
+        return  PyTuple_New(0);
```
Comment (vstinner) [inline Objects/typeobject.c]:
```suggestion
        return PyTuple_New(0);
```

**Review by vstinner — APPROVED** [review]
LGTM

**Comment by serhiy-storchaka** [comment]
Thank you for your review, Victor. I applied all your suggestions.

**Pull request merged**